### DISCUSSION TITLE
v3.0.0-alpha9.1

### DISCUSSION THREAD

**Discussion by meatballhat** [urfave/cli#1985] [Announcements]
## What's Changed
* Fix variadic args by @joshfrench in https://github.com/urfave/cli/pull/1865
* Update: Remove jfrog scan by @dearchap in https://github.com/urfave/cli/pull/1870
* Fix:(issue_1860) Remove hidden flags from flag categories by @dearchap in https://github.com/urfave/cli/pull/1869
* chore: fix struct name by @kaifulee in https://github.com/urfave/cli/pull/1877
* include a newline between different categories by @abennett in https://github.com/urfave/cli/pull/1861
* Stop command aliases and flag aliases from being suggested in shell completion by @bartekpacia in https://github.com/urfave/cli/pull/1882
* Update to latest codecov action by @meatballhat in https://github.com/urfave/cli/pull/1898
* Replace Jaro-Winkler algorithm usage with an internal function by @yigithankarabulut in https://github.com/urfave/cli/pull/1893
* Make it possible to `source` the Zsh autocomplete script by @bartekpacia in https://github.com/urfave/cli/pull/1896
* Update GitHub Actions (bump versions, improve naming a bit) by @bartekpacia in https://github.com/urfave/cli/pull/1895
* Add gofumpt to CI by @bartekpacia in https://github.com/urfave/cli/pull/1899
* Update and cleanup readme by @bartekpacia in https://github.com/urfave/cli/pull/1901
* Include flag descriptions when using zsh completion by @Toalaah in https://github.com/urfave/cli/pull/1903
* improve readme a bit more by @bartekpacia in https://github.com/urfave/cli/pull/1906
* Run go mod tidy by @meatballhat in https://github.com/urfave/cli/pull/1907
* Target `stable` and `oldstable` Go version in CI by @Juneezee in https://github.com/urfave/cli/pull/1912
* Fix:(issue_1916) Fix subcommand flag completion by @dearchap in https://github.com/urfave/cli/pull/1918
* Fix typos by @deining in https://github.com/urfave/cli/pull/1923
* Fix:(issue_1925) Dont print default text for required flags by @dearchap in https://github.com/urfave/cli/pull/1935
* adjust tests for HideDefault by @Skeeve in https://github.com/urfave/cli/pull/1937
* fix: disable shell completion if double dash is included in arguments (v3) by @suzuki-shunsuke in https://github.com/urfave/cli/pull/1933
* docs: fix broken links by @suzuki-shunsuke in https://github.com/urfave/cli/pull/1940
* Fix:(issue_1930) Fix for invalid bool counts by @dearchap in https://github.com/urfave/cli/pull/1941
* chore(README): correct cli-altsrc link in README by @vimiix in https://github.com/urfave/cli/pull/1943
* Add File() by @6543 in https://github.com/urfave/cli/pull/1945
* feat(command): show global options in command help by @BlackHole1 in https://github.com/urfave/cli/pull/1949
* docs(flag): add `UseShortOptionHandling` description by @BlackHole1 in https://github.com/urfave/cli/pull/1953
* Feat:(issue_1958) Add support for multiple layouts to TimestampFlag by @horockey in https://github.com/urfave/cli/pull/1959
* fix(bool-with-inverse-flag): fix string printing of bool with inverse flag to properly show inverse prefix by @mahadzaryab1 in https://github.com/urfave/cli/pull/1966
* Fix:  Shell completions are broken when login shell is different than currently active shell #1884 by @imransolanki in https://github.com/urfave/cli/pull/1969
* Feat:(issue_1960) make defaults validation optional by @dearchap in https://github.com/urfave/cli/pull/1961
* feat: added HideHelpCommands to docs by @midirhee12 in https://github.com/urfave/cli/pull/1968
* Fix:(issue_1775) Add test for zero value structs by @dearchap in https://github.com/urfave/cli/pull/1972
* Fix:(issue_1951) Propagate hide help to subcommands by @dearchap in https://github.com/urfave/cli/pull/1952
* Dont check required flags until command action is executed by @dearchap in https://github.com/urfave/cli/pull/1975
* Fix: Make flags persistent by default by @dearchap in https://github.com/urfave/cli/pull/1977
* Update mkdocs and dependencies by @meatballhat in https://github.com/urfave/cli/pull/1978
* Update dependencies by @meatballhat in https://github.com/urfave/cli/pull/1983

## New Contributors
* @kaifulee made their first contribution in https://github.com/urfave/cli/pull/1877
* @abennett made their first contribution in https://github.com/urfave/cli/pull/1861
* @bartekpacia made their first contribution in https://github.com/urfave/cli/pull/1882
* @yigithankarabulut made their first contribution in https://github.com/urfave/cli/pull/1893
* @Toalaah made their first contribution in https://github.com/urfave/cli/pull/1903
* @deining made their first contribution in https://github.com/urfave/cli/pull/1923
* @Skeeve made their first contribution in https://github.com/urfave/cli/pull/1937
* @vimiix made their first contribution in https://github.com/urfave/cli/pull/1943
* @6543 made their first contribution in https://github.com/urfave/cli/pull/1945
* @horockey made their first contribution in https://github.com/urfave/cli/pull/1959
* @mahadzaryab1 made their first contribution in https://github.com/urfave/cli/pull/1966
* @imransolanki made their first contribution in https://github.com/urfave/cli/pull/1969
* @midirhee12 made their first contribution in https://github.com/urfave/cli/pull/1968

**Full Changelog**: https://github.com/urfave/cli/compare/v3.0.0-alpha9...v3.0.0-alpha9.1

<hr /><em>This discussion was created from the release <a href='https://github.com/urfave/cli/releases/tag/v3.0.0-alpha9.1'>v3.0.0-alpha9.1</a>.</em>

**Comment by meatballhat** [comment]
This version looks funky because `v3.0.0-alpha10` is less than `v3.0.0-alpha9`, meaning we should have been doing `v3.0.0-alpha.{N}` all along. Lesson learned!

**Comment by meatballhat** [comment]
Annoyingly, removing the `v3.0.0-alpha10` release also deleted the discussion 😐 . There was a whole conversation about renaming the tag because of the problem with not having a dot between `alpha` and the number.

**Reply by suzuki-shunsuke** [reply]
I restored comments from email. 😅
But I couldn't restore my comments from email because GitHub doesn't send email to me for my comments.
So I wrote my comments based on my memory.

[meatballhat](https://github.com/meatballhat)

> FYI The go module proxy doesn't like alpha10 😭
> 
> [swappy-20241013_091803.png (view on web)](https://github.com/user-attachments/assets/f17540da-24ee-4118-94ab-a2b448ddcc95)
> 
> I'm personally of the opinion that the current state of main isn't ready to be declared beta-quality, which is why I went with another alpha release. More than 9 alpha releases is a bit silly, though, and given this issue with go module proxy, I'm thinking that re-tagging v3.0.0-beta1 is probably the right move. Opinions wanted! ❤️

[bartekpacia](https://github.com/bartekpacia)

> I'm really surprised that alpha9 works but alpha10 doesn't. Isn't it a bug on Google's side? Maybe we should report it?
> I'd prefer to not declare "beta" now, I agree we're not ready for it.

[suzuki-shunsuke](https://github.com/suzuki-shunsuke)

> I think this follows semver.
> 3.0.0-alpha9 > 3.0.0-alpha10 because `Identifiers with letters or hyphens are compared lexically in ASCII sort order.`
> If we add `.` between `alpha` and number, we can increment number over 10. `alpha.10`
> https://semver.org/
> > Precedence for two pre-release versions with the same major, minor, and patch version MUST be determined by comparing each dot separated identifier from left to right until a difference is found as follows:
> > 1. Identifiers consisting of only digits are compared numerically.
> > 2. Identifiers with letters or hyphens are compared lexically in ASCII sort order.
> > 3. Numeric identifiers always have lower precedence than non-numeric identifiers.
> > 4. A larger set of pre-release fields has a higher precedence than a smaller set, if all of the preceding identifiers are equal.
> > Example: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0.

[suzuki-shunsuke](https://github.com/suzuki-shunsuke)

> I found the official document.
> https://go.dev/doc/modules/version-numbers
> > A released module is published with a version number in the semantic versioning model, as in the following illustration:

[suzuki-shunsuke](https://github.com/suzuki-shunsuke)

> I think beta.1 is good. Then we will be able to increment versions to over beta.10 without any problem.

[meatballhat](https://github.com/meatballhat)

> [suzuki-shunsuke](https://github.com/suzuki-shunsuke) [bartekpacia](https://github.com/bartekpacia) WDYT about switching to v3.0.0-alpha.10 and then re-tagging all of the other v3.0.0-alpha.{N} releases to point at the same refs? IIUC the go module proxy will still retain all of the references to v3.0.0-alpha{N} so nobody with existing module entries should break.

[bartekpacia](https://github.com/bartekpacia)

> That sounds good to me.

[suzuki-shunsuke](https://github.com/suzuki-shunsuke)

> Unfortunately `v3.0.0-alpha{N}` > `v3.0.0-alpha.{N}`, so I think v3.0.0-alpha.{N}` doesn't work well.

Eng Zer Jun

> > I think beta.1 is good. Then we will be able to increment versions to over beta.10 without any problem.
> 
> +1. I think we should follow https://go.dev/doc/modules/version-numbers.
> 
> Now the question is whether we are still able to use the alpha tag to publish new versions. The source code of Go module semver comparison logic is here: https://github.com/golang/mod/blob/v0.21.0/semver/semver.go#L110-L138
> 
> I tried the semver.Compare function in Go playground (link: https://go.dev/play/p/fOxOaD_iivf). It seems like our only option here is to use beta, unless we want to go for something like v3.0.0-alpha9.1
> 
> ```go
> package main
> 
> import (
> 	"fmt"
> 
> 	"golang.org/x/mod/semver"
> )
> 
> func main() {
> 	fmt.Println(semver.Compare("v3.0.0-alpha9", "v3.0.0-alpha10"))  // prints 1, means v3.0.0-alpha9 > v3.0.0-alpha10
> 	fmt.Println(semver.Compare("v3.0.0-alpha9", "v3.0.0-alpha.11")) // prints 1 too, means v3.0.0-alpha9 > v3.0.0-alpha.11
> 	fmt.Println(semver.Compare("v3.0.0-alpha9", "v3.0.0-beta.1"))   // prints -1, means v3.0.0-alpha9 < v3.0.0-beta.1
> 
> 	fmt.Println(semver.Compare("v3.0.0-alpha9", "v3.0.0-alpha9.1")) // prints -1
> }
> ```

[meatballhat](https://github.com/meatballhat)

> Given the options between v3.0.0-alpha9.1 and v3.0.0-beta.1, I'm in favor of v3.0.0-alpha9.1, as interesting as that looks 🤷🏼. I don't want to use beta until the [remaining list of must-have items for v3 are done](https://github.com/urfave/cli/issues/833). I'm happy to hear other opinions ❤️

[meatballhat](https://github.com/meatballhat)

> Another proposal that would mean we would keep approaching but not quite reaching alpha10: https://go.dev/play/p/oBi80hAaB-K. Ha ha just serious.
> 
> ```go
> package main
> 
> import (
> 	"fmt"
> 
> 	"golang.org/x/mod/semver"
> )
> 
> func main() {
> 	fmt.Println(semver.Compare("v3.0.0-alpha9", "v3.0.0-alpha9.9"))      // prints -1, means v3.0.0-alpha9 < v3.0.0-alpha9.9
> 	fmt.Println(semver.Compare("v3.0.0-alpha9.9", "v3.0.0-alpha9.99"))   // prints -1, means v3.0.0-alpha9.9 < v3.0.0-alpha9.99
> 	fmt.Println(semver.Compare("v3.0.0-alpha9.99", "v3.0.0-alpha9.999")) // prints -1, means v3.0.0-alpha9.99 < v3.0.0-alpha9.999
> }
> ```

[suzuki-shunsuke](https://github.com/suzuki-shunsuke)

> I think v3.0.0-alpha9.1 is good if we don't want to release a beta version yet.

[meatballhat](https://github.com/meatballhat)

> Works for me! Thank you 🙇🏼

**Reply by meatballhat** [reply]
@suzuki-shunsuke Amazing! 🤩
